baseline-network-linux-hld.md

```diff
@@ -0,0 +1,250 @@
+April 2024.
+
+The initial Kernel version target could be 5.15 (minimum kernel version of
+Yocto) or 6.6 (the latest LTS version of the Linux kernel). The kernel patch
+version will be a moving target, and regularly updated (ideally weekly).
+
+Since there will be many out-of-tree drivers for the platform support, required
+API updates should be done via
+[coccinelle semantic patches](https://coccinelle.gitlabpages.inria.fr/website/).
+This should ensure that the amount of work needed for keeping drivers building
+and working with newer kernels stays manageable.
+
+The semantic patches will also help vendors quickly updating their drivers when
+submitting new platforms based on older releases.
+
+## Milestones
+
+- M1:
+  - initial platform support repository with hand-selected/-crafted platform
+    support (>= 1 platform)
+  - initial machine definitions
+  - initial layer with kernel with defconfig, and package for platform support
+  - initial kernel 5.15?
+  - ramdisk image bootable on selected platform(s)
+- M2:
+  - refined kernel configuration via kernel-meta
+  - import of various platforms via scripts from SONiC
+  - initial ONIE NOS installer
+- M3:
+  - import missing platforms from ONL (convert them to PDDF?)
+  - import missing platforms from DentOS (convert them to PDDF?)
+  - kernel 6.6 support(?)
+  - coccinelle semantic patches for upgrading drivers to support Linux 6.6 APIs
+  - ONLP adapter(?)
+  - extensible ONIE NOS installer
+- M4:
+  - get accepted as
+    [official yocto layer](https://www.yoctoproject.org/development/yocto-project-compatible-layers/)
+  - are we fine with LTS kernels, or do we need to support non-LTS ones?
+  - figure out long-term maintenance strategy
+
+## Possible Future Extensions
+
+### ASIC / dataplane support
+
+While providing ASIC / dataplane support is out of scope for this project,
+a follow up project could provide appropriate SAI or similar packaging.
+
+### Common userspace API / utilities
+
+Currently there are several competing ways of providing access to sensor data
+and other values. To avoid forcing users to know which one is used for their
+current platform, it would be nice to have a generic/common utilities for
+querying the data, hiding the implementation details.
+
+### Conformance testing or DIAG utilities
+
+To allow verifying that various drivers and libraries work, it would be good
+to have some frameworks or testing utilities for OEMs to ensure that everything
+works from the API side.
+
+### DentOS as a layer extending the base platform support layer
+
+The base platform support will be already provided by the base platform layer
+(i.e. custom hardware drivers, platform initialization, etc).
+
+Since DentOS uses a customized version of the Prestera driver that contains
+several features not included in the upstream driver, the custom Prestera
+driver will be provided as an external kernel module package, while the
+in-kernel driver will be disabled.
```

Comment:
If possible, selecting between the out-of-tree and mainline drivers via different Yocto machines (or just a config option) would allow easy comparison and testing of the different feature sets. That would in turn make improving the mainline driver easier.

Reply:
A feature should be (hopefully) easy to do and would make switching between both easier than switching between different machines (AFAICT).